Tools/build/generate_sbom.py

```diff
@@ -118,70 +124,211 @@ def filter_gitignored_paths(paths: list[str]) -> list[str]:
     return sorted([line.split()[-1] for line in git_check_ignore_lines if line.startswith("::")])
 
 
-def discover_pip_sbom_package(sbom_data: dict[str, typing.Any]) -> None:
-    """pip is a part of a packaging ecosystem (Python, surprise!) so it's actually
-    automatable to discover the metadata we need like the version and checksums
-    so let's do that on behalf of our friends at the PyPA.
+def fetch_package_metadata_from_pypi(project: str, version: str, filename: str | None = None) -> tuple[str, str] | None:
     """
-    global PACKAGE_TO_FILES
+    Fetches the SHA256 checksum and download location from PyPI.
+    If we're given a filename then we match with that, otherwise we use wheels.
+    """
+    # Get pip's download location from PyPI. Check that the checksum is correct too.
+    try:
+        raw_text = urlopen(f"https://pypi.org/pypi/{project}/{version}/json").read()
+        release_metadata = json.loads(raw_text)
+        url: dict[str, typing.Any]
+
+        # Look for a matching artifact filename and then check
+        # its remote checksum to the local one.
+        for url in release_metadata["urls"]:
+            # pip can only use Python-only dependencies, so there's
+            # no risk of picking the 'incorrect' wheel here.
+            if (
+                (filename is None and url["packagetype"] == "bdist_wheel")
+                or (filename is not None and url["filename"] == filename)
+            ):
+                break
+        else:
+            raise ValueError(f"No matching filename on PyPI for '{filename}'")
+
+        # Successfully found the download URL for the matching artifact.
+        download_url = url["url"]
+        checksum_sha256 = url["digests"]["sha256"]
+        return download_url, checksum_sha256
+
+    except (OSError, ValueError) as e:
+        # Fail if we're running in CI where we should have an internet connection.
+        error_if(
+            "CI" in os.environ,
+            f"Couldn't fetch metadata for project '{project}' from PyPI: {e}"
+        )
+        return None
+
+
+def find_ensurepip_pip_wheel() -> pathlib.Path | None:
+    """Try to find the pip wheel bundled in ensurepip. If missing return None"""
 
     ensurepip_bundled_dir = CPYTHON_ROOT_DIR / "Lib/ensurepip/_bundled"
-    pip_wheels = []
 
-    # Find the hopefully one pip wheel in the bundled directory.
+    pip_wheels = []
     for wheel_filename in os.listdir(ensurepip_bundled_dir):
```

Comment:
@befeleme This assumes that the `ensurepip/_bundled` directory is still around, maybe that's a bad assumption?

Reply:
I'll make this more resilient in case other projects delete the whole directory.

Reply:
Done in https://github.com/python/cpython/pull/114450/commits/1947f23bfba448839929dcacf29ded1ff4bc0f91